pkg/tracing/otlp/otlp.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package otlp
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	_ "google.golang.org/grpc/encoding/gzip"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	TracingClientGRPC string = "grpc"
+	TracingClientHTTP string = "http"
+)
+
+// NewOTELTracer returns an OTLP exporter based tracer.
+func NewTracerProvider(ctx context.Context, logger log.Logger, conf []byte) (*tracesdk.TracerProvider, error) {
+	config := Config{}
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, err
+	}
+
+	var exporter *otlptrace.Exporter
+	var err error
+	switch config.ClientType {
+	case TracingClientHTTP:
+		options := traceHTTPOptions(config)
+
+		client := otlptracehttp.NewClient(options...)
+		exporter, err = otlptrace.New(ctx, client)
+		if err != nil {
+			return nil, err
+		}
+
+	case TracingClientGRPC:
+		options := traceGRPCOptions(config)
+		client := otlptracegrpc.NewClient(options...)
+		exporter, err = otlptrace.New(ctx, client)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, errors.New("otlp: invalid client type. Only 'http' and 'grpc' are accepted. ")
+	}
+
+	tp := newTraceProvider(exporter)
+
+	return tp, nil
+}
+
+func newTraceProvider(exporter *otlptrace.Exporter) *tracesdk.TracerProvider {
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithBatcher(exporter),
+		tracesdk.WithResource(resource.NewWithAttributes(
```

Comment:
Should we also use `resource.New` here to obtain the resource to detect some resources automatically? Same as it is done in other exporters. For reference: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/README.md